wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Create/Logic/GroupConversationCreationCoordinator.swift

```diff
@@ -0,0 +1,218 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+
+protocol GroupConversationCreator {
+    func insertGroupConversation(
+        moc: NSManagedObjectContext,
+        participants: [ZMUser],
+        name: String?,
+        team: Team?,
+        allowGuests: Bool,
+        allowServices: Bool,
+        readReceipts: Bool,
+        messageProtocol: MessageProtocol) -> ZMConversation?
+}
+
+class DefaultGroupConversationCreator: GroupConversationCreator {
+    func insertGroupConversation(
+        moc: NSManagedObjectContext,
+        participants: [ZMUser],
+        name: String?,
+        team: Team?,
+        allowGuests: Bool,
+        allowServices: Bool,
+        readReceipts: Bool,
+        messageProtocol: MessageProtocol) -> ZMConversation? {
+            return ZMConversation.insertGroupConversation(
+                moc: moc,
+                participants: participants,
+                name: name,
+                team: team,
+                allowGuests: allowGuests,
+                allowServices: allowServices,
+                readReceipts: readReceipts,
+                messageProtocol: messageProtocol
+            )
+        }
+}
+
+class GroupConversationCreationCoordinator {
+    private enum State {
+        case ready
+        case initializing
+        case initialized(eventHandler: (GroupConversationCreationEvent) -> Void)
+        case beginingToCreateConversation(eventHandler: (GroupConversationCreationEvent) -> Void)
+        case creatingConversation(conversation: ZMConversation, eventHandler: (GroupConversationCreationEvent) -> Void)
+        case finalizing
+    }
```

Comment:
Is there a strong reason to have the `eventHandler` as an associated value of these enum cases ? Why not just assign it to a class property?

Reply:
This way I do not have to double-check states as only the states that should trigger handler contains it.
I didn't want handlers to be triggered when they shouldn't so it ensures that behaviour.

Reply:
I can see how we benefit from that but I'm not sure it justifies the extra complexity when we could check the state when accessing the `eventHandler`, if it was made into a computed property perhaps. Anyhow, it works the way it is now so I don't think it's necessary to change it. I'll leave this up to you